torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
 
 AliasDb::~AliasDb() = default;
 
+// Structure that keeps tracks of all writes at a high level. When alias
+// analysis is done, this will be used to construct a more efficient WriteIndex.
+struct AliasDb::WriteRegistry {
+    void registerWrite(const Value* v, Node* n) {
+      writes_.emplace(n, v);
+    }
+    void registerWriteToAllContained(const Value* v, Node* n) {
+      containedWrites_.emplace(n, v);
+    }
+    void registerWriteToAllWildcards(Node* n) {
+      writesToAllWildcards_.insert(n);
+    }
+    std::unordered_map<Node*, const Value*> writes_;
+    std::unordered_map<Node*, const Value*> containedWrites_;
+    std::unordered_set<Node*> writesToAllWildcards_;
+};
+
 AliasDb::AliasDb(std::shared_ptr<Graph> graph, bool isFrozen)
-    : graph_(std::move(graph)), isFrozen_(isFrozen) {
-  memoryDAG_ = torch::make_unique<MemoryDAG>();
+    : graph_(std::move(graph)),
+      isFrozen_(isFrozen),
+      memoryDAGBuilder_(std::make_unique<MemoryDAGBuilder>()),
+      writeRegistry_(std::make_unique<AliasDb::WriteRegistry>()) {
   analyze(graph_);
+  memoryDAG_ = memoryDAGBuilder_->build();
+
+  // null out the builder to make accessing it a hard error
+  memoryDAGBuilder_ = nullptr;
+
+  memoryDAG_->setWildcards(
+      wildcards_, elementMap_, [&](const Value* v) -> Element* {
+        return getWildcard(v->type());
+      });
+
+  // Initialize the write index
+  writeIndex_ = TWriteIndex();
+  // dereference to make operator[] less ugly.
+  auto& writeIndex = *writeIndex_;
+
+  // build the write index
+  for (const auto& write : writeRegistry_->writes_) {
+    auto it = elementMap_.find(write.second);
+    TORCH_INTERNAL_ASSERT(
+        it != elementMap_.end(), "Tried to write to value not in MemoryDAG");
+    const auto& writtenMemoryLocations =
+        memoryDAG_->getMemoryLocations(it->second);
+    writeIndex[write.first] |= writtenMemoryLocations;
+  }
+
+  for (const auto& write : writeRegistry_->containedWrites_) {
+    auto elem = elementMap_.at(write.second);
+    MemoryLocations mem_locations;
+    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
+    for (unsigned loc : mem_locations) {
+      auto contained_elem = memoryDAG_->fromIndex(loc);
+      // we only register writes on memory locations
+      if (contained_elem->pointsTo.empty()) {
+        writeIndex[write.first].set(contained_elem->index);
+      }
+    }
+  }
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
```

Comment:
I'm not sure why we have `writesToAllWildcards`. It's only called on `aten::wait`, where we could instead just `analyzeConservative`.

Reply:
the comment explains why. We cannot just be conservative and write to all inputs, because we don't know what the forked inputs are.

Reply:
If we analyzed `aten::fork` and `aten::wait` with `analyzeConservative` that would be correct, and would not require the above special casing leaking into memory dag

Reply:
```
f : Future[None] = fork(a, b, c)  // mutates a
wait(f)
use(a)
```
it is illegal to reorder `wait` after `use`. `analyzeConservative` will only register a write to `f`, not `a`.

This can probably be improved. But I don't want to change behavior in this PR if possible.

Reply:
`analyzeConservative` marks a write to all contained elements:
```
    auto elem = elementMap_.at(input);
    registerWrite(input, node);
    MemoryLocations mem_locations;
    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
    for (unsigned loc : mem_locations) {
      auto contained_elem = memoryDAG_->fromIndex(loc);
      // we only register writes on memory locations
      if (contained_elem->pointsTo.empty()) {
        registerWrite(contained_elem, node);
      }
    }
```